File: DBClientQt/DBClient/person.cpp
#include "person.h"

#include <QVariant>
#include <QDebug>
#include <QCryptographicHash>

using std::endl;
//QTextStream in(stdin);

Person::Person()
{
	values["persLogin"] = "[\\d\\w_]{1,30}";
	values["persName"] = "[\\d\\w]{1,30}";
	values["access"] = "[AMRS]{1}";
	values["phone"] = "[\\d]{1,11}";
	values["passport"] = "[\\d]{10}";
	values["persPassword"] = "[\\d\\w]{5,}";

	types["persLogin"] = true;
	types["persName"] = true;
	types["access"] = true;
	types["phone"] = true;
	types["passport"] = true;
	types["persPassword"] = true;

	replaces["persLogin"] = false;
	replaces["persName"] = true;
	replaces["access"] = false;
	replaces["phone"] = false;
	replaces["passport"] = false;
	replaces["persPassword"] = false;
}

Person::Person(const Person &pers)
{
	this->name = pers.name;
	this->access = pers.access;
	this->id = pers.id;
	this->phone = pers.phone;
	this->passport = pers.passport;
	this->login = pers.login;
}

bool Person::IsManager(QSqlDatabase db, int id)
	{
		QString query = "SELECT access FROM dbo.PersonsView WHERE id = " + QString::number(id);
		QSqlQuery sql(db);
		if (!sql.exec(query))
		{			
			return false;
		}
		sql.next();
		QString result = sql.value("access").toString();		
		return result == "M";
	}

QVector<Person> Person::GetPersons(const QSqlDatabase &db)
{
	QSqlQuery sql(db);
	if (!sql.exec("SELECT id, persLogin, persName, passport, access, phone FROM dbo.PersonsView"))
	{
		return QVector<Person>();
	}

	QVector<Person> persons;
	while(sql.next())
	{
		Person pers = Person();

		pers.access = sql.value("access").toString();
		pers.login = sql.value("persLogin").toString();
		pers.name = sql.value("persName").toString();
		pers.phone = sql.value("phone").toString();
		pers.passport = sql.value("passport").toString();
		pers.id = sql.value("id").toInt();

		persons.append(pers);
	}

	return persons;
}

Person Person::Auth(const QSqlDatabase &db, const QString &login, const QString &password)
{
	QSqlQuery sql(db);
	Person p;
	if (!sql.exec("SELECT id, persName, passport, phone, access, persPassword, persLogin \
				  FROM dbo.PersonsView \
				  WHERE persLogin LIKE '" + login + "'"))
	{
		p.id = -1;
		return p;
	}

	if (!sql.next())
	{
		p.id = -1;
		return p;
	}

	QString passwordRec = sql.value("persPassword").toString();
	QString res = QString(QCryptographicHash::hash(password.toUtf8(), QCryptographicHash::Md5).toHex());

	if (!res.contains(passwordRec))
	{
		p.id = -2;
		return p;
	}

	Person pers = Person();
	pers.access = sql.value("access").toString();
	pers.login = sql.value("persLogin").toString();
	pers.name = sql.value("persName").toString();
	pers.phone = sql.value("phone").toString();
	pers.passport = sql.value("passport").toString();
	pers.id = sql.value("id").toInt();
	return pers;
}

bool Person::ChangePerson(QSqlDatabase db, QString passport, QMap<QString, QString> params)
{
	QString query = "UPDATE dbo.PersonsView SET ";

	if (params.contains("persPassword"))
	{
		params["persPassword"] = QString(QCryptographicHash::hash(params["persPassword"].toUtf8(), QCryptographicHash::Md5).toHex());
	}

	QString additing = Service::IsCorretParam(params, Person().values, Person().types, Person().replaces);
	if (additing.trimmed().length() == 0)
	{
		return false;
	}

	query.append(additing + " WHERE passport = '" + passport + "'");

	QSqlQuery sql(db);
	return sql.exec(query);
}

bool Person::InsertPerson(QSqlDatabase db, QMap<QString, QString> params)
{
// add persons persLogin=asd123 persPassword=8126849 persName=jhsad passport=1234567899 phone=214 access=R
	QString query = "EXEC dbo.InsertNewPerson ";
	QString additing = Service::IsCorretParam(params, Person().values, Person().types, Person().replaces);
	if (additing.length() == 0 || !Service::contains_fields(params, Person().values.keys()))
	{
		qDebug() << "Ошибка при вводе параметров!";
		return false;
	}
	if (params.contains("persPassword"))
	{
		params["persPassword"] = QString(QCryptographicHash::hash(params["persPassword"].toUtf8(), QCryptographicHash::Md5).toHex());
	}
	QSqlQuery sql(db);
	return sql.exec(query + "'" + params["persName"] + "', '" +
		params["passport"] + "', '" + params["phone"] + "', '" + params["access"] +
		"', '" + params["persLogin"] + "', '" + params["persPassword"] + "', 0");
}

bool Person::DeletePerson(QSqlDatabase db, QString passport)
{
	QSqlQuery sql(db);
	return sql.exec("DELETE FROM dbo.PersonsView WHERE passport = '" + passport + "'");
}

QString Person::GetPersonId(QSqlDatabase db, QString passport)
{
	QSqlQuery sql(db);
	if (!sql.exec("SELECT id FROM dbo.PersonsView WHERE passport LIKE '" + passport + "'"))
	{
		qDebug() << "Ошибка выполнения запроса!";
		return "";
	}
	if (!sql.first())
	{
		qDebug() << "Пользователь не найден!";
		return "";
	}
	return QString::number(sql.value("id").toInt());
}

void Person::PrintPersons(const QVector<Person> &persons)
{
	qDebug() << "=============================================================";

	for(int i = 0; i < persons.size(); ++i)
	{
		Person p = Person(persons[i]);
		qDebug() << "#" << i << "\nИмя: " << p.getName() << "\nЛогин: " << p.getLogin() << "\nДолжность: "
				 << p.getAccess() << "\nПаспортные данные: " << p.getPassport() << "\nНомер телефона: " << p.getPhone() << endl;
	}
	qDebug() << "=============================================================";
	qDebug() << endl;
}

File: DBClientPhp/Person.php
<?php

include_once("sql_function.php");

if (!function_exists(sql_get_person_on_id))
{
	function sql_get_person_on_id($connection, $id)
	{
		$query = "SELECT id, persName, passport, phone, access, persLogin, persPassword FROM [fifaDataBase].[dbo].[PersonsView] WHERE id = ".$id;
		if (!$queryResult = sql_exec_query($connection, $query))
		{
			return false;
		}
		odbc_fetch_row($queryResult);

		return 	[ 0 => [
				'id' => odbc_result($queryResult, "id"), 
				'persName' => odbc_result($queryResult, "persName"), 
				'passport' => odbc_result($queryResult, "passport"), 
				'phone' => odbc_result($queryResult, "phone"), 
				'access' => odbc_result($queryResult, "access"), 
				'persLogin' => odbc_result($queryResult, "persLogin"), 
				] ];
	}
}

if (!function_exists(sql_auth))
{
	function sql_auth($login, $password, $connection)
	{
		$query = "SELECT id, persName, passport, phone, access, persPassword, persLogin FROM [fifaDataBase].[dbo].[PersonsView] WHERE persLogin LIKE '".$login."'  ";
		if (!$queryResult = sql_exec_query($connection, $query))
		{
			return -2;
		}
		odbc_fetch_row($queryResult);
		$pos = strpos(md5($password), odbc_result($queryResult, "persPassword"));
		if ($pos === false)
		{
			return -1;
		}

		return [ 'id' => odbc_result($queryResult, "id"), 'access' => odbc_result($queryResult, "access") ];
	}
}

if (!function_exists(sql_is_manager_query))
{
	function sql_is_manager_query($connection, $id)
	{
		$query = 'SELECT access FROM [fifaDataBase].[dbo].[PersonsView] WHERE id = '.$id;
		if (!$queryResult = sql_exec_query($connection, $query))
		{
			return false;
		}

		$i = 0;
		$result = '';
		while(odbc_fetch_row($queryResult))
		{
			$result = odbc_result($queryResult, "access");
		}
		odbc_free_result($queryResult);
		return $result;
	}
}

if (!function_exists(sql_select_persons))
{
	function sql_select_persons($connection)
	{
		$query = 'SELECT id, persName, passport, phone, access, persLogin FROM [fifaDataBase].[dbo].[PersonsView]';
		if (!$queryResult = sql_exec_query($connection, $query))
		{
			return false;
		}

		$i = 0;
		$result = [];
		while(odbc_fetch_row($queryResult))
		{
			$result[$i] = [
				'id' => odbc_result($queryResult, "id"), 
				'persName' => odbc_result($queryResult, "persName"), 
				'passport' => odbc_result($queryResult, "passport"), 
				'phone' => odbc_result($queryResult, "phone"), 
				'access' => odbc_result($queryResult, "access"), 
				'persLogin' => odbc_result($queryResult, "persLogin")
				];
			$i++;
		}
		odbc_free_result($queryResult);
		return $result;
	}
}

if (!function_exists(sql_insert_person))
{
	function sql_insert_person($connection, $array)
	{
		$query = "EXEC [fifaDataBase].[dbo].[InsertNewPerson] ";
		
		$fields = ['persName' => ['/[\d\w]{1,30}/', 30, true, true], 'persLogin' => ['/[\d\w]{1,30}/', 30, true, false], 'access' => ['/[AMSR]/', 1, true, false], 'phone' => ['/[\d]{1,11}/', 11, true, false], 'passport' => ['/[0-9]{10}/', 10, true, false], 'persPassword' => ['/[\d\w]{5,}/', 512, true, true]];
		
		if (get_update_fileds($array, $fields) === false || !contains_fields($array, $fields))
		{
			return false;
		}

		$query = $query."'".$array['persName']."', '".$array['passport']."', '".$array['phone']."', '".$array['access']."', '".$array['persLogin']."', '".md5($array['persPassword'])."', 0";

		return sql_exec_prepare_query($connection, $query);
	}
}

if (!function_exists(delete_person))
{
	function delete_person($connection, $id)
	{
		$query = 'DELETE FROM [fifaDataBase].[dbo].[PersonsView] WHERE id = '.$id;
		return sql_exec_prepare_query($connection, $query);
	}
}

if (!function_exists(sql_change_person))
{
	function sql_change_person($connection, $id, $array)
	{
		$query = "UPDATE [fifaDataBase].[dbo].[PersonsView] SET ";
		
		$fields = ['persName' => ['/[\d\w]{1,30}/', 30, true, true], 'persLogin' => ['/[\d\w]{1,30}/', 30, true, false], 'access' => ['/[AMSR]/', 1, true, false], 'phone' => ['/[\d]{1,11}/', 11, true, false], 'passport' => ['/[0-9]{10}/', 10, true, false], 'persPassword' => ['/[\d\w]{5,}/', 512, true, true]];
		if (array_key_exists('persPassword', $array))
		{
			if (strlen(trim(preg_replace($fields['persPassword'][0], " ", $array['persPassword']))) != 0 || strlen($array['persPassword']) > 30 || strlen($array['persPassword']) < 5)
			{
				echo "Error of enter password!\n";
				return false;
			}
			$array['persPassword'] = md5($array['persPassword']);
		}
		$additingQuery = get_update_fileds($array, $fields);
		if ($additingQuery === false)
		{
			return false;
		}
		$query = $query.$additingQuery."WHERE id = ".$id;

		return sql_exec_prepare_query($connection, $query);
	}
}

if (!function_exists(print_selected_persons))
{
	function print_selected_persons($array)
	{
		printf("%'=103s \n%-5s%-32s%-12s%-13s%-9s%-32s \n%'=103s \n", " ", "id", "persName", "passport", "phone", "access", "persLogin", " ");
		foreach ($array as $key => $value) 
		{
			printf("%-5s%-32s%-12s%-13s%-9s%-32s \n" , $value['id'], $value['persName'], $value['passport'], $value['phone'], $value['access'], $value['persLogin']);
		}
		printf("%'=103s \n", " ");
	}
}


?>